driver/src/main/java/org/neo4j/driver/internal/BaseDriver.java

```diff
@@ -43,4 +54,68 @@ public boolean isEncrypted()
         return securityPlan.requiresEncryption();
     }
 
+    public <T> T transact( RetryLogic logic, AccessMode mode, Function<Transaction, T> work )
+            throws NotCommittedException, ServiceUnavailableException
+    {
+        int remaining = logic.attempts();
+        while ( remaining > 0 )
+        {
+            try ( Session session = session( mode ) )
+            {
+                boolean failed = false;
+                Transaction tx = session.beginTransaction();
+                try {
+                    T result = work.apply( tx );
+                    tx.success();
+                    return result;
+                }
+                catch ( SessionExpiredException e )
+                {
+                    failed = true;
+                    tx.failure();
+                    remaining -= 1;
+                }
+                finally
+                {
+                    if ( failed )
+                    {
+                        try
+                        {
+                            tx.close();
+                        }
+                        catch ( Exception ex )
+                        {
+                            // ignore errors if we've already failed as
+                            // we already know this connection is problematic
+                        }
+                    }
+                    else
+                    {
+                        tx.close();
+                    }
+                }
+            }
+            try
+            {
+                Thread.sleep( logic.pauseMillis() );
+            }
+            catch ( InterruptedException e )
+            {
+                throw new NotCommittedException( format( "Interrupted after %d attempts", logic.attempts() - remaining ) );
+            }
+        }
+        throw new NotCommittedException( format( "Unable to commit transaction after %d attempts", logic.attempts() ) );
+    }
```

Comment:
I feel like the code in here is more complex than it needs to be, but this can be addressed later.